cmd/lotus-seal-worker/sub.go

```diff
@@ -0,0 +1,214 @@
+package main
+
+import (
+	"context"
+	"gopkg.in/cheggaaa/pb.v1"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/lib/sectorbuilder"
+)
+
+type worker struct {
+	api           api.StorageMiner
+	minerEndpoint string
+	repo          string
+	auth          http.Header
+
+	sb *sectorbuilder.SectorBuilder
+}
+
+func acceptJobs(ctx context.Context, api api.StorageMiner, endpoint string, auth http.Header, repo string) error {
+	act, err := api.ActorAddress(ctx)
+	if err != nil {
+		return err
+	}
+	ssize, err := api.ActorSectorSize(ctx, act)
+	if err != nil {
+		return err
+	}
+
+	sb, err := sectorbuilder.NewStandalone(&sectorbuilder.Config{
+		SectorSize:    ssize,
+		Miner:         act,
+		WorkerThreads: 1,
+		CacheDir:      filepath.Join(repo, "cache"),
+		SealedDir:     filepath.Join(repo, "sealed"),
+		StagedDir:     filepath.Join(repo, "staged"),
+		MetadataDir:   filepath.Join(repo, "meta"),
+	})
+	if err != nil {
+		return err
+	}
+
+	w := &worker{
+		api:           api,
+		minerEndpoint: endpoint,
+		auth:          auth,
+		repo:          repo,
+		sb:            sb,
+	}
+
+	tasks, err := api.WorkerQueue(ctx)
+	if err != nil {
+		return err
+	}
+
+	for task := range tasks {
+		log.Infof("New task: %d, sector %d, action: %d", task.TaskID, task.SectorID, task.Type)
+
+		res := w.processTask(ctx, task)
+
+		log.Infof("Task %d done, err: %s", task.TaskID, res.Err)
+
+		if err := api.WorkerDone(ctx, task.TaskID, res); err != nil {
+			log.Error(err)
+		}
+	}
+
+	log.Warn("acceptJobs exit")
+	return nil
+}
+
+func (w *worker) processTask(ctx context.Context, task sectorbuilder.WorkerTask) sectorbuilder.SealRes {
+	switch task.Type {
+	case sectorbuilder.WorkerPreCommit:
+	case sectorbuilder.WorkerCommit:
+	default:
+		return errRes(xerrors.Errorf("unknown task type %d", task.Type))
+	}
+
+	if err := w.fetchSector(task.SectorID, task.Type); err != nil {
```

Comment:
does this mean it fetches the data over http twice per sector?

Reply:
Yes

Reply:
(can optimize later)